Yank/mpi.py

```diff
@@ -0,0 +1,421 @@
+#!/usr/bin/env python
+
+# ==============================================================================
+# MODULE DOCSTRING
+# ==============================================================================
+
+"""Utilities to run on MPI.
+
+Provide functions and decorators that simplify running the same code on
+multiple nodes. One benefit is that serial and parallel code is exactly
+the same.
+
+Global variables
+----------------
+disable_mpi : bool
+    Set this to True to force running serially.
+
+Routines
+--------
+get_mpicomm
+    Automatically detect and configure MPI execution and return an
+    MPI communicator.
+run_single_node
+    Run a task on a single node.
+on_single_node
+    Decorator version of run_single_node.
+distribute
+    Map a task on a sequence of arguments on all the nodes.
+delay_termination
+    A context manager to delay the response to termination signals.
+delayed_termination
+    A decorator version of delay_termination.
+
+"""
+
+
+# ==============================================================================
+# GLOBAL IMPORTS
+# ==============================================================================
+
+import os
+import sys
+import signal
+import logging
+from contextlib import contextmanager
+
+# TODO drop this when we drop Python 2 support
+from openmoltools.utils import wraps_py2
+
+logger = logging.getLogger(__name__)
+
+# ==============================================================================
+# GLOBAL VARIABLES
+# ==============================================================================
+
+disable_mpi = False
+
+
+# ==============================================================================
+# MAIN FUNCTIONS
+# ==============================================================================
+
+def get_mpicomm():
+    """Retrieve the MPI communicator for this execution.
+
+    The function automatically detects if the program runs on MPI by checking
+    specific environment variables set by various MPI implementations. On
+    first execution, it modifies sys.excepthook and register a handler for
+    SIGINT, SIGTERM, SIGABRT to call Abort() to correctly terminate all
+    processes.
+
+    Returns
+    -------
+    mpicomm : mpi4py communicator or None
+        The communicator for this node, None if the program doesn't run
+        with MPI.
+
+    """
+    # If MPI execution is forcefully disabled, return None.
+    if disable_mpi:
+        return None
+
+    # If we have already initialized MPI, return the cached MPI communicator.
+    if get_mpicomm._is_initialized:
```

Comment:
Yeah, this is the only way in Python to have some sort of "static" function variable. There is indeed an inefficiency bug actually now that you point it out, since I need to set `get_mpicomm._is_initialized = True` also in the if-block below this that handles the non-MPI case.